README.md
### dbConnect_LE

Object oriented database engine for WebsiteBaker and LEPTON CMS

#### Requirements

* minimum PHP 5.2.x
* using [WebsiteBaker] [1]
* _or_ using [LEPTON CMS] [2]

#### Installation

* download the actual [dbConnectLE_x.xx.zip] [3] installation archive
* in CMS backend select the file from "Add-ons" -> "Modules" -> "Install module"

#### First Steps

You can access the *dbConnect_LE* database engine from your own addons:

    if (!class_exists('dbconnectle')) {
      require_once WB_PATH.'/modules/dbconnect_le/include.php';
    }

Now you can use dbConnect to extend your own database object:

    class dbArticles extends dbConnectLE {    
      
      const FIELD_ID = 'article_id';
      const FIELD_NUMBER = 'article_number';
      const FIELD_NAME = 'article_name';
      const FIELD_TIMESTAMP = 'article_timestamp';
    
      private $createTable = false;
    
      /**
       * Constructor
       *
       * @param $createTable boolean
       */
      public function __construct($createTable = false) {
        $this->createTable = $createTable;
        parent::__construct();
        $this->setTableName('mod_articles');
        $this->addFieldDefinition(self::FIELD_ID, "INT(11) NOT NULL AUTO_INCREMENT", true);
        $this->addFieldDefinition(self::FIELD_NUMBER, "VARCHAR(64) NOT NULL DEFAULT ''");
        $this->addFieldDefinition(self::FIELD_NAME, "VARCHAR(64) NOT NULL DEFAULT ''");
        $this->addFieldDefinition(self::FIELD_TIMESTAMP, "TIMESTAMP");
        $this->setIndexFields(array(self::FIELD_NAME));
        $this->checkFieldDefinitions();
        // create table
        if ($this->createTable) {
          if (!$this->sqlTableExists()) {
            if (!$this->sqlCreateTable()) {
              $this->setError(sprintf('[%s - %s] %s', __METHOD__, __LINE__, $this->getError()));
            }
          }
        }
        // set timezone
        date_default_timezone_set(CFG_TIME_ZONE);
      } // __construct()
    
    } // class dbArticles

dbConnect_LE will do a lot of jobs in a smart way.

    $table = new dbArcticles(true);
    
Will create a new database object and will also create the needed MySQL table if it not exists.

    $article = array();
    $SQL = sprintf("SELECT * FROM %s WHERE `%s`='%d'",
      $dbOfferArticles->getTableName(),
      dbOfferArticles::FIELD_ID,
      $id);
    if (!$dbOfferArticles->sqlExec($SQL, $article)) {
      $this->setError(sprintf('[%s - %s] %s', __METHOD__, __LINE__, $dbOfferArticles->getError()));
      return false;
    }
    print_r($articles);
    
Will select the article with the specified $id from the table.

Please have a look to any add-on of phpManufaktur for WebsiteBaker or LEPTON CMS - you will find a lot of examples.          

[1]: http://websitebaker2.org "WebsiteBaker Content Management System"
[2]: http://lepton-cms.org "LEPTON CMS"
[3]: https://github.com/phpManufaktur/dbConnect_LE/downloads


CHANGELOG
### dbConnect_LE

0.71 - 2013-11-23

* general program maintenance, no functional changes

0.70 - 2012-06-15

* fixed a WebsiteBaker 2.8.x compatibility problem

0.69 - 2012-05-13

* changed version checking for WebsiteBaker

0.68 - 2012-05-25

* internal use only, not published

0.67 - 2012-05-23

* moved repository from SVN to GitHub
* added: README.md, CHANGELOG and LICENSE
* removed: info.txt

0.66 - 2011-12-29

* removed: the old WebsiteBaker search feature is no longer supported

0.65 - 2011-07-20

* added: setAutoIncrement() to set start value of TABLE AUTO_INCREMENT
* grant compatibillity for LEPTON CMS

0.64 - 2011-02-18

* clean up code and fixed some minor problems

0.63 - 2010-11-21

* added: function sqlFoundRows() - executing FOUND_ROWS()

0.62 - 2010-10-24

* fixed: class dbConnectLE called some language constants of dbConnect

0.61 - 2010-10-17

* added global search feature for WebsiteBaker modules and addons
* clean up code and create documentation with phpDocumentor

0.60 - 2010-10-16

* first release - ported from dbConnect Release 0.36


precheck.php
<?php

/**
 * dbConnect_LE
 *
 * @author Ralf Hertsch <ralf.hertsch@phpmanufaktur.de>
 * @link http://phpmanufaktur.de
 * @copyright 2008 - 2013
 * @license MIT License (MIT) http://www.opensource.org/licenses/MIT
 */

// include class.secure.php to protect this file and the whole CMS!
if (defined('WB_PATH')) {
    if (defined('LEPTON_VERSION')) include (WB_PATH . '/framework/class.secure.php');
} else {
    $oneback = "../";
    $root = $oneback;
    $level = 1;
    while (($level < 10) && (! file_exists($root . '/framework/class.secure.php'))) {
        $root .= $oneback;
        $level += 1;
    }
    if (file_exists($root . '/framework/class.secure.php')) {
        include ($root . '/framework/class.secure.php');
    } else {
        trigger_error(sprintf("[ <b>%s</b> ] Can't include class.secure.php!", $_SERVER['SCRIPT_NAME']), E_USER_ERROR);
    }
}
// end include class.secure.php

// Checking Requirements
if (!defined('LEPTON_VERSION'))
  $PRECHECK ['WB_VERSION'] = array ('VERSION' => '2.8', 'OPERATOR' => '>=' );
$PRECHECK ['PHP_VERSION'] = array ('VERSION' => '5.2.0', 'OPERATOR' => '>=' );

global $database;
$sql = "SELECT * FROM " . TABLE_PREFIX . "settings WHERE name='default_charset'";
$result = $database->query($sql);
if ($result) {
	$data = $result->fetchRow();
	($data ['value'] == 'utf-8') ? $status = true : $status = false;
	$PRECHECK ['CUSTOM_CHECKS'] = array ('Default Charset' => array ('REQUIRED' => 'utf-8', 'ACTUAL' => $data ['value'], 'STATUS' => $status ) );
}


languages/DE.php
<?php

/**
 * dbConnect_LE
 *
 * @author Ralf Hertsch <ralf.hertsch@phpmanufaktur.de>
 * @link http://phpmanufaktur.de
 * @copyright 2008 - 2013
 * @license MIT License (MIT) http://www.opensource.org/licenses/MIT
 */

// include class.secure.php to protect this file and the whole CMS!
if (defined('WB_PATH')) {
    if (defined('LEPTON_VERSION')) include (WB_PATH . '/framework/class.secure.php');
} else {
    $oneback = "../";
    $root = $oneback;
    $level = 1;
    while (($level < 10) && (! file_exists($root . '/framework/class.secure.php'))) {
        $root .= $oneback;
        $level += 1;
    }
    if (file_exists($root . '/framework/class.secure.php')) {
        include ($root . '/framework/class.secure.php');
    } else {
        trigger_error(sprintf("[ <b>%s</b> ] Can't include class.secure.php!", $_SERVER['SCRIPT_NAME']), E_USER_ERROR);
    }
}
// end include class.secure.php

if ('á' != "\xc3\xa1") {
    // important: language files must be saved as UTF-8 (without BOM)
    trigger_error('The language file <b>/modules/'.dirname(basename(__FILE__)).'/languages/'.
	    basename(__FILE__).'</b> is damaged, it must be saved <b>UTF-8</b> encoded!', E_USER_ERROR);
}

define ( 'dbc_error_database_not_connected', 'Die Datenbank ist nicht verbunden!' );
define ( 'dbc_error_fieldDefinitionsNotChecked', 'Bitte rufen Sie zunächst die Funktion<strong>checkFieldDefinitions()</strong> auf, bevor Sie Datenbankabfragen durchführen!' );
define ( 'dbc_error_emptyTableName', 'Der Tabellenname ist leer, bitte definieren Sie zunächst einen Namen fuer die Tabelle!' );
define ( 'dbc_error_noFieldDefinitions', 'Die Felddefinitionen fehlen, legen Sie die zunächst die Felder fuer die Datenbank fest!' );
define ( 'dbc_error_noPrimaryKey', 'Der Primäre Schlüssel ist nicht definiert!' );
define ( 'dbc_error_execQuery', 'Fehler bei der SQL Abfrage: <b>%s</b>' );
define ( 'dbc_error_feature_not_supported', 'Diese Funktion wird von dbConnectLE nicht unterstützt!' );
define ( 'dbc_error_recordEmpty', 'Der übergebene Datensatz ist leer, es ist nichts zu tun...' );
define ( 'dbc_error_csv_file_no_handle', 'Fuer die CSV Datei <b>%s</b> konnte kein Handle erzeugt werden.' );
define ( 'dbc_error_csv_no_keys', 'In der CSV Datei <b>%s</b> wurden keine Spaltenüberschriften gefunden!<br />Fügen Sie Spaltenüberschriften in die CSV ein oder setzen Sie den Schalter <b>$has_header=false</b>.' );
define ( 'dbc_error_csv_file_put', 'Fehler beim Schreiben der CSV Datei <b>%s</b>.' );


languages/EN.php
<?php

/**
 * dbConnect_LE
 *
 * @author Ralf Hertsch <ralf.hertsch@phpmanufaktur.de>
 * @link http://phpmanufaktur.de
 * @copyright 2008 - 2013
 * @license MIT License (MIT) http://www.opensource.org/licenses/MIT
 */

// try to include LEPTON class.secure.php to protect this file and the whole CMS!
if (defined('WB_PATH')) {
	if (defined('LEPTON_VERSION')) include(WB_PATH.'/framework/class.secure.php');
} elseif (file_exists($_SERVER['DOCUMENT_ROOT'].'/framework/class.secure.php')) {
	include($_SERVER['DOCUMENT_ROOT'].'/framework/class.secure.php');
} else {
	$subs = explode('/', dirname($_SERVER['SCRIPT_NAME']));	$dir = $_SERVER['DOCUMENT_ROOT'];
	$inc = false;
	foreach ($subs as $sub) {
		if (empty($sub)) continue; $dir .= '/'.$sub;
		if (file_exists($dir.'/framework/class.secure.php')) {
			include($dir.'/framework/class.secure.php'); $inc = true;	break;
		}
	}
	if (!$inc) trigger_error(sprintf("[ <b>%s</b> ] Can't include LEPTON class.secure.php!", $_SERVER['SCRIPT_NAME']), E_USER_ERROR);
}
// end include LEPTON class.secure.php

// Be aware to save this file in UTF-8 (no BOM) to keep special chars!

define ( 'dbc_error_database_not_connected', 'Database is not connected!' );
define ( 'dbc_error_fieldDefinitionsNotChecked', 'Please call function <strong>checkFieldDefinitions()</strong> before executing SQL Queries.' );
define ( 'dbc_error_emptyTableName', 'Empty table name, please define table name at first.' );
define ( 'dbc_error_noFieldDefinitions', 'No field definitions, please define fields for the database.' );
define ( 'dbc_error_noPrimaryKey', 'Undefined Primary Key, please define a primary key for the table.' );
define ( 'dbc_error_execQuery', 'Error while executing SQL Query: <b>%s</b>' );
define ( 'dbc_error_feature_not_supported', 'This function is not supported by dbConnectLE' );
define ( 'dbc_error_recordEmpty', 'The delivered record is empty (no data to process).' );
define ( 'dbc_error_csv_file_no_handle', 'Can\'t create handle for the CSV file <b>%s</b>!' );
define ( 'dbc_error_csv_no_keys', 'Missing column headers for the CSV file <b>%s</b>! Insert headers or set switch <b>$has_header=false</b>.' );
define ( 'dbc_error_csv_file_put', 'Error writing CSV the file <b>%s</b>.' );

?>

languages/index.php
<?php

/**
 * dbConnect_LE
 *
 * @author Ralf Hertsch <ralf.hertsch@phpmanufaktur.de>
 * @link http://phpmanufaktur.de
 * @copyright 2008 - 2013
 * @license MIT License (MIT) http://www.opensource.org/licenses/MIT
 */

// include class.secure.php to protect this file and the whole CMS!
if (defined('WB_PATH')) {
    if (defined('LEPTON_VERSION')) include (WB_PATH . '/framework/class.secure.php');
} else {
    $oneback = "../";
    $root = $oneback;
    $level = 1;
    while (($level < 10) && (! file_exists($root . '/framework/class.secure.php'))) {
        $root .= $oneback;
        $level += 1;
    }
    if (file_exists($root . '/framework/class.secure.php')) {
        include ($root . '/framework/class.secure.php');
    } else {
        trigger_error(sprintf("[ <b>%s</b> ] Can't include class.secure.php!", $_SERVER['SCRIPT_NAME']), E_USER_ERROR);
    }
}
// end include class.secure.php


info.php
<?php

/**
 * dbConnect_LE
 *
 * @author Ralf Hertsch <ralf.hertsch@phpmanufaktur.de>
 * @link http://phpmanufaktur.de
 * @copyright 2008 - 2013
 * @license MIT License (MIT) http://www.opensource.org/licenses/MIT
 */

// include class.secure.php to protect this file and the whole CMS!
if (defined('WB_PATH')) {
    if (defined('LEPTON_VERSION')) include (WB_PATH . '/framework/class.secure.php');
} else {
    $oneback = "../";
    $root = $oneback;
    $level = 1;
    while (($level < 10) && (! file_exists($root . '/framework/class.secure.php'))) {
        $root .= $oneback;
        $level += 1;
    }
    if (file_exists($root . '/framework/class.secure.php')) {
        include ($root . '/framework/class.secure.php');
    } else {
        trigger_error(sprintf("[ <b>%s</b> ] Can't include class.secure.php!", $_SERVER['SCRIPT_NAME']), E_USER_ERROR);
    }
}
// end include class.secure.php

$module_directory = 'dbconnect_le';
$module_name = 'dbConnectLE';
$module_function = (defined('LEPTON_VERSION') || defined('CAT_VERSION')) ? 'library' : 'snippet';
$module_version = '0.71';
$module_platform = '2.8.x';
$module_author = 'Ralf Hertsch, Berlin (Germany)';
$module_license = 'MIT License (MIT)';
$module_description = 'Objectoriented MySQL interface for usage with WebsiteBaker, LEPTON CMS and BlackCat CMS';
$module_home = 'https://phpmanufaktur.de/';
$module_guid = 'A6697E72-9FC0-433C-8536-2112F6E6D687';
